File: codeProject/1-aiOpsPoc/multi-agent-aiops-main/multi-agent-aiops-main/golang/internal/model/events.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type Severity string

const (
	SeverityCritical Severity = "critical"
	SeverityHigh     Severity = "high"
	SeverityMedium   Severity = "medium"
	SeverityLow      Severity = "low"
	SeverityInfo     Severity = "info"
)

type HealLevel string

const (
	HealLevelL0Auto   HealLevel = "L0"
	HealLevelL1Semi   HealLevel = "L1"
	HealLevelL2Manual HealLevel = "L2"
)

// AlertEvent 监控告警事件
type AlertEvent struct {
	EventID       string            `json:"event_id"`
	Timestamp     time.Time         `json:"timestamp"`
	AlertName     string            `json:"alert_name"`
	Severity      Severity          `json:"severity"`
	Source        string            `json:"source"`
	TargetService string            `json:"target_service"`
	MetricName    string            `json:"metric_name"`
	MetricValue   float64           `json:"metric_value"`
	Threshold     float64           `json:"threshold"`
	Description   string            `json:"description"`
	Labels        map[string]string `json:"labels"`
}

// RCAResult 根因分析结果
type RCAResult struct {
	AlertEventID     string                   `json:"alert_event_id"`
	RootCause        string                   `json:"root_cause"`
	Confidence       float64                  `json:"confidence"`
	AffectedServices []string                 `json:"affected_services"`
	Evidence         []map[string]interface{} `json:"evidence"`
	SuggestedActions []string                 `json:"suggested_actions"`
}

// HealAction 自愈操作
type HealAction struct {
	RCAEventID      string                 `json:"rca_event_id"`
	HealLevel       HealLevel              `json:"heal_level"`
	ActionType      string                 `json:"action_type"`
	ActionParams    map[string]interface{} `json:"action_params"`
	TargetService   string                 `json:"target_service"`
	EstimatedImpact string                 `json:"estimated_impact"`
	BlastRadius     float64                `json:"blast_radius"`
	DryRunResult    string                 `json:"dry_run_result"`
	ExecutionResult string                 `json:"execution_result"`
}

// ChangeDecision 变更审批决策
type ChangeDecision struct {
	HealEventID    string  `json:"heal_event_id"`
	RiskScore      float64 `json:"risk_score"`
	ApprovalStatus string  `json:"approval_status"`
	Approver       string  `json:"approver"`
	Reason         string  `json:"reason"`
}

// IncidentState 故障事件全局状态
type IncidentState struct {
	IncidentID     string                 `json:"incident_id"`
	CreatedAt      time.Time              `json:"created_at"`
	UpdatedAt      time.Time              `json:"updated_at"`
	Status         string                 `json:"status"`
	AlertEvent     *AlertEvent            `json:"alert_event,omitempty"`
	RCAResult      *RCAResult             `json:"rca_result,omitempty"`
	HealAction     *HealAction            `json:"heal_action,omitempty"`
	ChangeDecision *ChangeDecision        `json:"change_decision,omitempty"`
	CurrentAgent   string                 `json:"current_agent"`
	RetryCount     int                    `json:"retry_count"`
	ErrorMessage   string                 `json:"error_message,omitempty"`
	Metadata       map[string]interface{} `json:"metadata"`
}

func NewIncidentState() *IncidentState {
	return &IncidentState{
		IncidentID: uuid.New().String(),
		CreatedAt:  time.Now(),
		UpdatedAt:  time.Now(),
		Status:     "open",
		Metadata:   make(map[string]interface{}),
	}
}

File: codeProject/1-aiOpsPoc/multi-agent-aiops-main/multi-agent-aiops-main/python/models/events.py
"""事件驱动架构的核心数据模型

所有 Agent 间通信都基于这些事件模型，通过 Kafka 事件总线传递。
"""

from datetime import datetime
from enum import Enum
from typing import Any, Optional
from pydantic import BaseModel, Field
import uuid


class Severity(str, Enum):
    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"
    INFO = "info"


class AlertStatus(str, Enum):
    FIRING = "firing"
    RESOLVED = "resolved"
    SUPPRESSED = "suppressed"


class AgentType(str, Enum):
    MONITOR = "monitor"
    RCA = "rca"
    HEAL = "heal"
    CHANGE = "change"
    ORCHESTRATOR = "orchestrator"


class EventType(str, Enum):
    ALERT_FIRED = "alert.fired"
    ALERT_RESOLVED = "alert.resolved"
    RCA_STARTED = "rca.started"
    RCA_COMPLETED = "rca.completed"
    HEAL_PROPOSED = "heal.proposed"
    HEAL_EXECUTING = "heal.executing"
    HEAL_COMPLETED = "heal.completed"
    HEAL_FAILED = "heal.failed"
    CHANGE_REQUESTED = "change.requested"
    CHANGE_APPROVED = "change.approved"
    CHANGE_REJECTED = "change.rejected"


class HealLevel(str, Enum):
    """自愈分级策略"""
    L0_AUTO = "L0"       # 全自动，无需确认
    L1_SEMI = "L1"       # 半自动，需确认
    L2_MANUAL = "L2"     # 人工介入


class BaseEvent(BaseModel):
    event_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    event_type: EventType
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    source_agent: AgentType
    correlation_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    metadata: dict[str, Any] = Field(default_factory=dict)


class AlertEvent(BaseEvent):
    """监控告警 Agent 产生的告警事件"""
    event_type: EventType = EventType.ALERT_FIRED
    source_agent: AgentType = AgentType.MONITOR

    alert_name: str
    severity: Severity
    status: AlertStatus = AlertStatus.FIRING
    source: str                        # 告警源（如 prometheus, loki）
    target_service: str                # 受影响服务
    metric_name: Optional[str] = None
    metric_value: Optional[float] = None
    threshold: Optional[float] = None
    description: str = ""
    labels: dict[str, str] = Field(default_factory=dict)


class RCAEvent(BaseEvent):
    """根因分析 Agent 产生的分析结果事件"""
    event_type: EventType = EventType.RCA_COMPLETED
    source_agent: AgentType = AgentType.RCA

    alert_event_id: str                # 关联的原始告警
    root_cause: str                    # 根因描述
    confidence: float                  # 置信度 0-1
    affected_services: list[str]       # 受影响的服务链路
    evidence: list[dict[str, Any]]     # 证据链
    suggested_actions: list[str]       # 建议修复动作


class HealEvent(BaseEvent):
    """故障自愈 Agent 产生的修复事件"""
    event_type: EventType = EventType.HEAL_PROPOSED
    source_agent: AgentType = AgentType.HEAL

    rca_event_id: str                  # 关联的根因分析
    heal_level: HealLevel
    action_type: str                   # 修复类型（restart_pod, scale_up, rollback...）
    action_params: dict[str, Any]      # 修复参数
    target_service: str
    estimated_impact: str              # 预估影响
    blast_radius: float                # 爆炸半径 (0-1)
    dry_run_result: Optional[str] = None
    execution_result: Optional[str] = None


class ChangeEvent(BaseEvent):
    """变更审批 Agent 产生的审批事件"""
    event_type: EventType = EventType.CHANGE_REQUESTED
    source_agent: AgentType = AgentType.CHANGE

    heal_event_id: str
    risk_score: float                  # 风险评分 0-1
    approval_status: str = "pending"   # pending / approved / rejected
    approver: str = ""                 # 审批人（auto 表示自动审批）
    reason: str = ""
    audit_log: list[dict[str, Any]] = Field(default_factory=list)


class IncidentState(BaseModel):
    """一次完整故障处理的全局状态（流经 Orchestrator）"""
    incident_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    status: str = "open"               # open / investigating / healing / resolved

    alert_event: Optional[AlertEvent] = None
    rca_event: Optional[RCAEvent] = None
    heal_event: Optional[HealEvent] = None
    change_event: Optional[ChangeEvent] = None

    current_agent: AgentType = AgentType.MONITOR
    retry_count: int = 0
    error_message: Optional[str] = None
